homeassistant/components/mqtt/__init__.py

```diff
@@ -1000,9 +1000,37 @@ def discovery_callback(payload):
 class MqttEntityDeviceInfo(Entity):
     """Mixin used for mqtt platforms that support the device registry."""
 
-    def __init__(self, device_config: Optional[ConfigType]) -> None:
+    def __init__(self, device_config: Optional[ConfigType],
+                 config_entry=None) -> None:
         """Initialize the device mixin."""
         self._device_config = device_config
+        self._config_entry = config_entry
+
+    async def device_info_discovery_update(self, config: dict):
+        """Handle updated discovery message."""
+        self._device_config = config.get(CONF_DEVICE)
+        device_registry = await \
+            self.hass.helpers.device_registry.async_get_registry()
+        config_entry_id = self._config_entry.entry_id
```

Comment:
Check if `self._config_entry` is None before. If the user tries to configure the device registry through YAML, this line will raise an `AttributeError`

Reply:
I thought device registry only works through discovery as documented: https://www.home-assistant.io/components/switch.mqtt/#device
Can you give an example of when this errors so I can test and verify your suggestion?

Reply:
Example is in previous comment.

Previously, this would not do anything in the case of yaml config. But now this will raise an attributeerror. We can do better than that. Let's print a proper warning/error

Reply:
But function `device_info_discovery_update` is only called if mqtt discovery message is received, how can this happen for a device configured trough yaml? An example of exactly what yaml triggers an error would be helpful.

Reply:
I think this is fixed now.